File: 0834/solution.ts
function sumOfDistancesInTree(n: number, edges: number[][]): number[] {
  if (n == 1) {
    return [0];
  }
  const distances = new Array<number>(n).fill(0);
  const counts = new Array<number>(n).fill(1);
  const tree = new Map<number, Array<number>>();
  for (const [x, y] of edges) {
    tree.set(x, tree.get(x) || new Array<number>());
    tree.set(y, tree.get(y) || new Array<number>());
    tree.get(x)!.push(y);
    tree.get(y)!.push(x);
  }
  const getCounts = (node: number, parent: number): number => {
    let dist = 0;
    for (const child of tree.get(node)!) {
      if (child != parent) {
        dist += getCounts(child, node) + counts[child];
        counts[node] += counts[child];
      }
    }
    return dist;
  };

  const getDistances = (node: number, parent: number): void => {
    for (const child of tree.get(node)!) {
      if (child != parent) {
        distances[child] =
          distances[node] - counts[child] + (n - counts[child]);
        getDistances(child, node);
      }
    }
  };

  distances[0] = getCounts(0, -1);
  getDistances(0, -1);
  return distances;
}

File: 0834/solution.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Solution {
    Map<Integer, List<Integer>> tree;
    int[] counts;
    int[] distances;
    int N;

    public int[] sumOfDistancesInTree(int n, int[][] edges) {
        if (n == 1) {
            return new int[] { 0 };
        }
        N = n;
        counts = new int[n];
        distances = new int[n];
        tree = new HashMap<>();
        for (int[] edge : edges) {
            int x = edge[0];
            int y = edge[1];
            tree.putIfAbsent(x, new ArrayList<Integer>());
            tree.putIfAbsent(y, new ArrayList<Integer>());
            tree.get(x).add(y);
            tree.get(y).add(x);
        }
        distances[0] = getCounts(0, -1);
        getDistances(0, -1);
        return distances;
    }

    private int getCounts(int node, int parent) {
        int distance = 0;
        for (int child : tree.get(node)) {
            if (child != parent) {
                distance += getCounts(child, node) + counts[child];
                counts[node] += counts[child];
            }
        }
        counts[node]++;
        return distance;
    }

    private void getDistances(int node, int parent) {
        for (int child : tree.get(node)) {
            if (child != parent) {
                distances[child] = distances[node] - counts[child] + (N - counts[child]);
                getDistances(child, node);
            }
        }
    }
}

// time O(n)
// space O(n)